Project Insights endpoints and dashboards

EddieLF opened the pull request:
This is a WIP draft PR for implementing a "Projects Summary" web endpoint. The goal is to provide users with summary stats for one or more projects they are a member of. 

So far, the stats captured here are the number of families, participants, samples, sequencing groups, and completed CRAMs, grouped by dataset AND sequencing type. 

**TODO**: 
So, in the web layer I've put each of the relevant queries into a function, but this is basically just clutter right now, the queries could exist as variables and there would be no difference.

Right now, the code iterates through the input projects and sequencing types and evaluate each of the 5 queries for all combinations of project and sequencing type, adding the result to the response array on each loop. 
This seems suuper inefficient, and I think the more obvious way to do this would be to evaluate all 5 queries for all input projects & sequencing types, _and then_ filter the results of this complete query down into distinct `ProjectsSummaryInternal` objects. 

So, could we, upon calling the `get_projects_summary(projects, sequencing_types)` function, evaluate the 5 queries for all projects and seq types (with a `fetch_all`), then subset the results of this by dataset and seq type once we get to the main loop? This way the 5 queries would only be evaluated once each. Just not sure how to do this right now.

Additional stats to add include the number of sequencing groups in the latest ES-Index and in the latest joint-call.

EddieLF commented:
An implementation of the Seqr projects dashboard.

- Use new ProjectAPI endpoint `get_project_insights_stats(projects, sequencing_types)` to retrieve counts of families, participants, samples, sequencing groups, CRAMs, the latest joint call analysis, SNV ES-Index analysis, and SV ES-Index analysis - **for each dataset + sequencing type combination**
- Render these as rows in a table located at seqr/stats, with the option to select which projects and sequencing types to display

TODO:
- Refactor the naming of the new files and classes, etc. Currently a mess of "Project Insights", "Insights Stats", "Insights Details", "Seqr Stats"... will choose one to decide on
- Get information about the dataset analyses (AnnotateDataset, SNV Es-Index, SV ES-Index) more generically rather than with fixed fields in the data model
- Add more of this analysis information to the mouseover tooltips - e.g. Created at timestamp, Output path
- Add a "Detailed" view, which can provide sequencing group level granularity for the selected sequencing projects and types. Use `Virtuoso` to handle the rows without declaring explicit pagination.

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -0,0 +1,667 @@
+
+        response = []
+        for pid in projects:
+            project = await ptable.get_and_check_access_to_project_for_id(self.author, pid, readonly=True)
```
A call to check access to multiple projects (save this async function being in a loop)

illusional reviewed (changes requested):
As discussed on zoom

illusional commented on api/routes/project_insights.py:
```diff
@@ -0,0 +1,52 @@
+    response_model=list[ProjectInsightsSummary],
+)
+async def get_project_insights_summary(
+    project_ids: list[int] = None,
```
Return `project_names: list[str]`, and then do map to IDs (which checks project access)

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -189,26 +192,37 @@ async def _stats_sequencing_groups_query(
+        self,
+        project_ids: list[ProjectId],
+        sequencing_types: list[str],
+    ) -> defaultdict[tuple[ProjectId, str, str], list[int]]:
```
I think, there's a SequencingGroupInternalId or something

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -220,18 +234,25 @@ async def _stats_crams_query(
+
+    async def _latest_annotate_dataset_by_project_id_and_seq_type(
+        self, project_ids: list[ProjectId], sequencing_types: list[str]
+    ) -> defaultdict[tuple[ProjectId, str], dict[str, Any]]:
```
Because you use SequencingTechnology / Type so much, just declare an alias above to allow this to read easier

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+                    row["sequencing_platform"],
+                    row["sequencing_technology"],
+                )
+            ].append(row)
```
Beware returning row pointer here, it's not a dict. In this case, it carries a lot of extra fields, and you need to know the keys by looking the code. Consider returning a FamilyInternal object, as you can leverage typing to understand exactly what the response looks like

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+            latest_es_indices_by_project_id_and_seq_type_and_stage[
+                (row["project"], row["sequencing_type"], row["stage"])
+            ] = {
+                "id": row["id"],
```
Maybe create a namedtuple, or tiny dataclass for this, be your own friend for typing

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+                "projects": project_ids,
+            },
+        )
+        return {(row['project'], row['sequencing_group_id']): row for row in _query_results}
```
Same here, don't return a row, look for a type you can serialise

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+LEFT JOIN analysis_sequencing_group asg on asg.analysis_id=a.id
+INNER JOIN (
+    SELECT
+        asg.sequencing_group_id,
```
Return the a.id here, of the max completed (with order by).

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+    AND JSON_EXTRACT(meta, '$.stage') = 'Stripy'
+    GROUP BY asg.sequencing_group_id
+) max_timestamps ON asg.sequencing_group_id = max_timestamps.sequencing_group_id
+AND a.timestamp_completed = max_timestamps.max_timestamp
+WHERE a.project IN :projects
+AND a.type = 'web'
+AND JSON_EXTRACT(meta, '$.stage') = 'Stripy';
+        """
```
You don't need these conditions if you check on the inner query

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+    GROUP BY asg.sequencing_group_id
+) max_timestamps ON asg.sequencing_group_id = max_timestamps.sequencing_group_id
+AND a.timestamp_completed = max_timestamps.max_timestamp
+WHERE a.project IN :projects
```
Same treatement as above, but you can probably just ditch the inner query because the other fields are the same

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+                "projects": project_ids,
+            },
+        )
+        return {(row['project'], row['sequencing_group_id']): row for row in _query_results}
```
Pick a type here

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+    # Helper functions
+    async def _get_sequencing_groups_by_analysis_ids(
+        self, analysis_ids: list[int]
+    ) -> dict[int, list[int]]:
```
Looking for the SequencingGroupInternalId alias here

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+
+    def get_sg_web_report_links(
+        self,
+        sequencing_group_stripy_reports,
```
Type annotations brother

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+        if stripy_report:
+            report_links['stripy'] = {
+                'stripy_url': (
+                    f'https://main-web.populationgenomics.org.au/'
```
Grab the gs path from the analysis-object, and add a transformation function. Easy to write tests for

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+            self._connection
+        ).get()
+
+        # Get all possible permutations of the projects, sequencing types, and sequencing technologies
```
```suggestion
        # Get all possible combinations of the projects, sequencing types, and sequencing technologies
```

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+            )
+
+            if sequencing_technology == "short-read":
+                latest_annotate_dataset = AnalysisStatsInternal(
```
This function is long, and nested makes it harder to read, consider moving this construction to it's own function

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+            self._connection
+        ).get()
+
+        # Get all possible permutations of the projects, sequencing types, sequencing platforms, and sequencing technologies
```
combinations

illusional commented on db/python/layers/project_insights.py:
```diff
@@ -301,367 +332,601 @@ async def _stats_es_indexes_query(
+                continue
+
+            if sequencing_technology == "short-read":
+                latest_annotate_dataset = AnalysisStatsInternal(
```
Same function complexity break up here

illusional commented on models/models/project_insights.py:
```diff
@@ -1,152 +1,160 @@
 @dataclass
 class ProjectInsightsDetailsInternal:
-    """Return class for the project insights details endpoint"""
+    """Return class for the seqr projects insights details endpoint"""
```
```suggestion
    """Return class for the projects insights details endpoint"""
```

illusional commented on web/src/pages/projectInsights/Details.tsx:
```diff
@@ -0,0 +1,239 @@
+const Details: React.FC = () => {
+    // Get the list of  projects from the project API
+    const [allData, setAllData] = useState<ProjectInsightsDetails[]>([])
+    const [projectNames, setProjectNames] = React.useState<string[]>([])
```
@dancoates, can you take a look at the typescript + react on this function, to see if you have suggestions on improving the code here

illusional commented on web/src/pages/projectInsights/DetailsTable.tsx:
```diff
@@ -0,0 +1,903 @@
+// DetailsTable.tsx
+import React, { useState } from 'react'
+import { ProjectInsightsDetails } from '../../sm-api'
+import { Button, Table, Icon } from 'semantic-ui-react'
```
Instead of the table here, look for the other defined Table in this function, you'll still want to import Table as SUITable so you can access the SUITable.Row / .Cell, etc. But our table definition has some handy dark mode, compact customisations

illusional commented on web/src/pages/projectInsights/DetailsTable.tsx:
```diff
@@ -0,0 +1,903 @@
+            <Table.Cell data-cell className="table-cell">
+                {details.sequencing_group_report_links?.stripy ? (
+                    <a
+                        href={`https://main-web.populationgenomics.org.au/${details.dataset}/stripy/${details.sequencing_group_id}.html`}
```
Can you return the URL from the server?

illusional commented on web/src/pages/projectInsights/DetailsTable.tsx:
```diff
@@ -0,0 +1,903 @@
+    const exportToFile = (format: 'csv' | 'tsv') => {
+        const headerCells = document.querySelectorAll('.ui.table thead tr th')
+        const headerData = Array.from(headerCells).map((cell) => cell.textContent)
+        const rows = document.querySelectorAll('.ui.table tbody tr')
```
Query with a specific table ID, rather than a generic selector

illusional commented on web/src/pages/projectInsights/DetailsTable.tsx:
```diff
@@ -0,0 +1,903 @@
+                <Button onClick={() => exportToFile('csv')}>Export to CSV</Button>
+                <Button onClick={() => exportToFile('tsv')}>Export to TSV</Button>
+            </div>
+            <Table sortable>
```
he internal table

illusional commented on web/src/pages/projectInsights/DetailsTable.tsx:
```diff
@@ -0,0 +1,903 @@
+                        >
+                            Seq Type
+                        </Table.HeaderCell>
+                        <Table.HeaderCell
```
1. Pull this out as it's own component
2. You have a clear, what's required to actually this header cell
3. `[{name: "platform", title: "Platform", selectedOptions: selectedSeqPlatforms}, ...]`

illusional commented on web/src/pages/projectInsights/Summary.tsx:
```diff
@@ -0,0 +1,116 @@
+    const [selectedSeqTypes, setSelectedSeqTypes] = React.useState<string[]>([])
+    const [selectedSeqTechnologies, setSelectedSeqTechnologies] = React.useState<string[]>([])
+
+    useEffect(() => {
```
You _could_, get allMyProjects, and then just have the seqr ones selected

illusional commented on web/src/pages/projectInsights/Summary.tsx:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    return (
+        <div>
```
```suggestion
        <>
```

illusional commented on web/src/pages/projectInsights/Summary.tsx:
```diff
@@ -0,0 +1,116 @@
+                selectedSeqTechnologies={selectedSeqTechnologies}
+                handleSelectionChange={handleSelectionChange}
+            ></SummaryTable>
+        </div>
```
```suggestion
        </>
```

illusional commented on web/src/pages/projectInsights/SummaryTable.tsx:
```diff
@@ -0,0 +1,733 @@
+            ? (summary.total_crams / summary.total_sequencing_groups) * 100
+            : 0
+
+    const percentageInJointCall =
+        summary.latest_annotate_dataset?.sg_count ?? 0 > 0
+            ? ((summary.latest_annotate_dataset?.sg_count ?? 0) / summary.total_sequencing_groups) *
+              100
+            : 0
```
```suggestion
    const latestAnnotateSgCount = summary.latest_annotate_dataset?.sg_count ?? 0
    const percentageInJointCall = latestAnnotateSgCount > 0
            ? (latestAnnotateSgCount / summary.total_sequencing_groups) * 100
            : 0
```

illusional commented on web/src/pages/projectInsights/SummaryTable.tsx:
```diff
@@ -0,0 +1,733 @@
+                <div style={{ display: 'flex', justifyContent: 'center' }}>
+                    <HtmlTooltip
+                        title={
+                            <p>
```
Remove the para tags, I think it'll be fine

illusional commented on web/src/pages/projectInsights/SummaryTable.tsx:
```diff
@@ -0,0 +1,733 @@
+                            </p>
+                        }
+                    >
+                        <div>{percentageAligned.toFixed(2)}%</div>
```
I don't think HTML loves divs being inside table cells

illusional commented on web/src/pages/projectInsights/SummaryTable.tsx:
```diff
@@ -0,0 +1,733 @@
+            <Table sortable>
+                <Table.Header>
+                    <Table.Row>
+                        <Table.HeaderCell
```
Simplify here

illusional reviewed (changes requested):
As discussed on call, looping in @dancoates for react / typescript stuff

EddieLF commented:
Hi @illusional - once again thank you for the last round of feedback. Once again requesting a re-review from you :) 🙏 
Please feel free to handball this to anyone else if it helps get it through. My goal right now is to have this in a satisfactory state to merge, allowing it to be more easily iterated on by future, smaller PRs. 

## Main changes since last review
- Use `getMyProjects` instead of `getSeqrProjects` when fetching available projects
- Seriously overhauled typing in `project_insights_layer.py`, now everything is very heavily typed
- Removed repetitive chunks of code in the layer, and reduced nesting / indentation by defining more functions - specifically for the `get_project_insights_details` and `get_project_insights_summary` functions at the bottom of the layer
- Better querying in the layer, using max timestamp and joining on analysis ID where required, so queries are much more readable. Also using `GROUP_CONCAT` when gathering many values in a grouping query.
- Removed all references to project ID on the user-side / front-end, only project names are used now
- Added "Insights" as a dropdown to Metamist webui navbar, with two options, "Summary" or "Details" which take you to `/insights/summary` and `/insights/details` respectively.
- Heavily refactored frontend to remove instances of highly repetitive code, especially as relates to table headers / columns.
  - Now, `HeaderCell.tsx` defines the possible column names, as well as the list of column names specifically for both the Summary and Details tables.
  - These columns for each table are exported as `summaryTableHeaderCellConfigs` and `detailsTableHeaderCellConfigs` - these are then passed around and used by the other parts of the module. This means the table columns are all defined in a centralised place and reused from there, making future updates significantly easier. 
- Refactored `ProjectAndSeqTypeSelector.tsx` to make the code here less repetitive & more re-usable, now project and sequencing type selections are both instances of the same `Selector` component. 
- Refactored `SummaryTable.tsx` and `DetailsTable.tsx`
  - These components now contain only functions and information to render cell content, handle sorting, and handle exporting.
- Refactored `Summary.tsx` and `Details.tsx`
  - These components now handle getting the initial data, storing the state for which filters have been applied (all stored in the `selectedValues` and `filteredData` state - no longer one state per column filter) and passing the data on to the `Table` components.
- Added dedicated `FilterData.tsx` component, to handle the filtering generically. Now when users click the filters on each header, the values shown are a combination of the already-applied filters, and the available values of that column with no filtering applied. Bit hard to explain, but we don't want values in the same column to drop out of the available filters immediately after filtering on a value, otherwise we can't select multiple values to filter on.
- Moved the `HtmlToolTip.tsx` into its own component in `/web/shared/utilities/`. 
- Created dedicated testing module for the Project Insights endpoints.


## TODO in future
- Add more tests!
- Add endpoints to GraphQL with filtering
- Improve layers logic, further reduce repetitive sections of code, optimise & generalise queries where possible
- Add more columns (specifically STRipy report outlier loci and report dates in the DetailsTable, SV calling status in DetailsTable, QC report links in SummaryTable)
- Generally improve UI and UX 
- Add new dashboards with more visualisations of the state of analyses in the various projects
- Transparently display which analysis (e.g. the es-indices, the annotated dataset) each sequencing group is a part of in the DetailsTable view
- Unify the Dashboards so they are accessed from the same page, with a toggle to display which dashboard table specifically, while keeping the filtering and project/seq type selection the same
- Add "export to csv-tsv" button for the SummaryTable dashboard

EddieLF commented:
<img width="1393" alt="Screenshot 2024-07-02 at 9 25 11 AM" src="https://github.com/populationgenomics/metamist/assets/34049565/564c68e4-eb8d-47ae-8a2a-15d04a963f32">
<img width="1374" alt="Screenshot 2024-07-02 at 9 26 09 AM" src="https://github.com/populationgenomics/metamist/assets/34049565/1aa6f973-97c0-4032-ab7e-b8b45a5bee3e">

illusional commented on web/src/pages/insights/DetailsTable.tsx:
```diff
@@ -0,0 +1,202 @@
+
+    const sortedData = React.useMemo(() => {
+        const data = [...filteredData]
+        data.sort((a, b) => {
```
Maybe consider using the lodash orderBy method, it allows multiple column sort, and returns the new sorted array for you. Though I think it uses the javascript sort operator

EddieLF replied:
lodash orderBy looks really well suited for this. We can pass the column names as an array, sort directions `'asc'`/`'desc'` array, and get the sorted array out. Seems really simple.

However it got a bit tricky in practice because the sorting logic is tied up in the semantic-ui-react `Table.HeaderCell` component, which has a `sorted` property that only allows the values `'ascending'` and `'descending'`. 

As far as I can tell, the sorting is mostly functional as it stands. I agree that using lodash orderBy will be more optimal, but I might leave that for the next PR if it's not too big a deal. Cheers

illusional commented on web/src/pages/insights/HeaderCell.tsx:
```diff
@@ -0,0 +1,245 @@
+    )
+}
+
+export const detailsTableHeaderCellConfigs: HeaderCellConfig[] = [
```
Naming conventions are [a bit over the place](https://stackoverflow.com/questions/40291766/naming-convention-for-const-object-keys-in-es6), but I'd probably suggest using screaming snake case, to suggest it's a global constant (and ditto below)

```suggestion
export const DETAILS_TABLE_HEADER_CELL_CONFIGS: HeaderCellConfig[] = [
```

illusional commented on web/src/pages/insights/SummaryTable.tsx:
```diff
@@ -0,0 +1,294 @@
+        summary.total_sequencing_groups > 0
+            ? (summary.total_crams / summary.total_sequencing_groups) * 100
+            : 0
+    const alignedCellTooltip = (
```
This is great to have it pulled out, you'd usually go a step further and make this it's own tiny component

EddieLF replied:
Pulled each of the "percentage" cells into their own individual components, e.g. `GetAlignedPercentageCell`, `GetAnnotatedDatasetPercentageCell`, etc. 

Now, the `GetSummaryCell` component is a lot simpler and to the point.
```tsx
const GetSummaryCell: React.FC<{
    summary: ProjectInsightsSummary
    columnKey: ColumnKey
    isDarkMode: boolean
}> = ({ summary, columnKey, isDarkMode }) => {
    switch (columnKey) {
        case 'aligned_percentage':
            return <GetAlignedPercentageCell summary={summary} isDarkMode={isDarkMode} />
        case 'annotated_dataset_percentage':
            return <GetAnnotatedDatasetPercentageCell summary={summary} isDarkMode={isDarkMode} />
        case 'snv_index_percentage':
            return <GetSnvIndexPercentageCell summary={summary} isDarkMode={isDarkMode} />
        case 'sv_index_percentage':
            return <GetSvIndexPercentageCell summary={summary} isDarkMode={isDarkMode} />
        default:
            return (
                <SUITable.Cell className="table-cell">
                    {summary[columnKey as keyof ProjectInsightsSummary]}
                </SUITable.Cell>
            )
    }
}
```

illusional commented on web/src/pages/insights/SummaryTable.tsx:
```diff
@@ -0,0 +1,294 @@
+    }
+}
+
+const getSummaryCell = (
```
This is a component, make sure to follow the standard component format + naming convention

illusional commented on web/src/pages/insights/SummaryTable.tsx:
```diff
@@ -0,0 +1,294 @@
+    const rowClassName = getRowClassName(summary.sequencing_type)
+    return (
+        <SUITable.Row
+            key={`${summary.dataset}-${summary.sequencing_type}-${summary.sequencing_technology}`}
```
FWIW, you only need a key inside a .map within a function, and the key just has to be neighbourly unique (within a component).

illusional commented on web/src/pages/insights/SummaryTable.tsx:
```diff
@@ -0,0 +1,294 @@
+                    <SUITable.Row>
+                        {summaryTableHeaderCellConfigs.map((config) => (
+                            <HeaderCell
+                                key={config.key}
```
EG: here you do need the component key, and it's neighbourly unique which is perf 👌

illusional commented on web/src/pages/insights/SummaryTable.tsx:
```diff
@@ -0,0 +1,294 @@
+                                key={config.key}
+                                config={config}
+                                sortDirection={
+                                    sortColumns.find((column) => column.column === config.key)
```
Nitpick: you could probably use a map, rather than an `O(n)` lookup for each column

illusional reviewed (approved):
💪 Huge effort @EddieLF, nothing critical from me, I think this is pretty good to go!
Great work pushing on that abstraction and getting this code nice and clean - this looks nice and easy to follow :)

If you're able to address each comment before you merge, but I don't feel like I need to review this again. You'll theoretically need a bumpversion commit because you've added routes, but practically if it's only used in typescript, those are bundled together, so less stressed.

EddieLF commented on db/python/layers/project_insights.py:
```diff
@@ -426,158 +682,105 @@ async def _details_stripy_reports(self, project_ids: list[ProjectId]) -> dict[Pr
         ptable = ProjectPermissionsTable(self._connection)
-        projects = await ptable.get_and_check_access_to_projects_for_ids(
-            user=self.author, project_ids=project_ids, readonly=True
+        projects = await ptable.get_and_check_access_to_projects_for_names(
```
Thanks for the heads up on this. 
Hey @dancoates, I just took a look through #829, what I am understanding is that a lot of the `ProjectPermissionsTable` methods like `get_and_check_access_to_project_for_id` and `get_and_check_access_to_projects_for_names` will be deprecated, and we should transition to using the method `get_projects_accessible_by_user` instead? 

If so, it should be fairly easy to adapt here. I will just have to add a check if the projects the user has access to fully overlap with the projects they're requesting in this function.

dancoates replied:
Hey @EddieLF the `get_and_check...` methods are still around, they've just moved from the ProjectPermissionsTable to the `connection` which should be available pretty much everywhere. They are also synchronous now as there is no need for a call to the db. You shouldn't need to use the `get_projects_accessible_by_user` method directly. 

https://github.com/populationgenomics/metamist/blob/fbc2897002829ca4c7f542ec855d42beaf8fddb6/db/python/connect.py#L133

I'm planning to write up some docs for this but feel free to reach out if anything isn't clear in the meantime.

EddieLF replied:
Ah gotcha, thanks for clearing that up.

EddieLF commented:
@illusional I've brought in the auth changes and latest updates from dev and implemented them into the project_insights db layer here, as well as addressing most of the changes you outlined above.

I know there's some funky codecov stuff happening, but I am happy to have this merged this into dev if there are no urgent issues to address! Thank you very much again for your time and patience :)
